Add division of self overlapping polygons

fercer opened the pull request:
## Description

Implementation of the division algorithm from Uddipan Mukherjee. (2014). Self-overlapping curves: Analysis and applications. Computer-Aided Design, 46, 227-232.
This algorithm allows to divide self-overlapping polygons into non self-overlapping polygons.
Currently, when drawing self-overlapping polygons, the overlapped regions are treated as holes.
This method allows to divide the polygon and draw or analyze each sub-polygon by separate.

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

fercer committed:
Added sopoly.py a module to sub-divide self-overlapping polygons

fercer committed:
Added a gallery example and tests for the divide_selfoverlapping function

pep8speaks commented:
Hello @fercer! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`skimage/measure/sopoly.py`](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py):

> [Line 96:78](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L96): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 266:72](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L266): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 342:69](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L342): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 360:35](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L360): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 361:35](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L361): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 362:35](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L362): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 465:46](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L465): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 475:19](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L475): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 786:72](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L786): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1276:25](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1276): [E121](https://duckduckgo.com/?q=pep8%20E121) continuation line under-indented for hanging indent
> [Line 1277:25](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1277): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1284:25](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1284): [E121](https://duckduckgo.com/?q=pep8%20E121) continuation line under-indented for hanging indent
> [Line 1285:25](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1285): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1290:41](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1290): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1294:41](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1294): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1326:61](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1326): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1558:63](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1558): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1560:63](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1560): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1561:33](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1561): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 1562:63](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1562): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1657:41](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1657): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1657:60](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1657): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1657:67](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1657): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1671:62](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1671): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1701:45](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1701): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1714:49](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1714): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1730:40](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1730): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1736:46](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/sopoly.py#L1736): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator

* In the file [`skimage/measure/tests/test_sopoly.py`](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/tests/test_sopoly.py):

> [Line 23:73](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/tests/test_sopoly.py#L23): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 24:67](https://github.com/scikit-image/scikit-image/blob/af662ff58fb2fd70eabc216a2f35316818328080/skimage/measure/tests/test_sopoly.py#L24): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','

mkcor commented on doc/examples/edges/plot_self_overlapping_polygon.py:
```diff
@@ -0,0 +1,60 @@
+
+This example shows how to divide a self-overlapping polygon into non
+self-overlapping sub polygons. Each sub-polygon can be drawn or analyzed by
+separate.
```
```suggestion
separately.
```

Can you please include the reference at the end of this intro? Syntax goes like: https://github.com/scikit-image/scikit-image/blob/b8ce4d0c2c65684b131870abfe60228136108d4e/doc/examples/segmentation/plot_mask_slic.py#L19-L22

mkcor commented on doc/examples/edges/plot_self_overlapping_polygon.py:
```diff
@@ -0,0 +1,60 @@
+import numpy as np
+import matplotlib.pyplot as plt
+
+from skimage.measure import divide_selfoverlapping
+from skimage.draw import polygon
```
I like importing submodules in the gallery examples (i.e., `from skimage import draw, measure`), so when the reader sees the function calls, they know that `divide_selfoverlapping` is a function `measure` etc. I think it brings some educational value.

fercer replied:
Absolutely! I updated the example to import the submodules and show that `divide_selfoverlapping` is inside `measure`.

mkcor commented on skimage/measure/tests/test_sopoly.py:
```diff
@@ -0,0 +1,122 @@
+                              [254, 265], [199, 269], [169, 248], [146, 277],
+                              [168, 307]])
+
+# A self-overlapping polygon with traversion direction given in counter
```
```suggestion
# A self-overlapping polygon with traversal direction given in counter
```

mkcor reviewed (commented):
Thank you so much for this contribution! Could you please add a (sub)title to each subplot or, alternatively, display them separately in different sections, so the reader can see what each image means? Thanks again!

fercer committed:
Apply suggestions from code review to fix style and documentation

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

fercer committed:
Updated example of self-overlapping polygons to add subtitles with the description of each plot

mkcor commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1789 @@
+
+
+def _shift_indices(a1, a2, n_vertices, ex_start=True, ex_end=True):
+    """ Shifts the position of the indices in a way that wraps them connecting
```
```suggestion
    """Shift the position of the indices in a way that wraps them connecting
```
cf. https://peps.python.org/pep-0257/#multi-line-docstrings

fercer replied:
Thank you again @mkcor for reviewing my code. I'll fix all the docstrings following the pep that you referenced.

mkcor commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1789 @@
+
+
+def _get_cut(a, vert_info, next_cut=True, same_ray=True, sign=0):
+    """ Looks for the next/previous valid vertex that can be used as cut
```
```suggestion
    """Look for the next/previous valid vertex that can be used as cut
```

mkcor commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1789 @@
+    ----------
+    a : int
+        An index of a vertex on the polygon used as reference to look for a cut
+        that satisfyes the given conditions.
```
```suggestion
        that satisfies the given conditions.
```

mkcor commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1789 @@
+
+
+def _check_adjacency(a1, a2, vert_info, left2right=True):
+    """ Check if the cut point a2 is adjacent to the cut a1 to the left/right.
```
```suggestion
    """Check whether cut point a2 is adjacent to cut point a1 to the left/right.
```

mkcor commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1789 @@
+    Returns
+    -------
+    is_adjacent : bool
+        If the indices a1 and a2 are adjacent when looked from left to right,
```
```suggestion
        Whether indices a1 and a2 are adjacent when looked from left to right,
```

mkcor reviewed (commented):
The CI build has failed. Are you be able to troubleshoot?

fercer committed:
Apply suggestions from code review on docstrings

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

fercer committed:
Updated docstrings on sopoly.py to follow PEP 8

fercer commented:
Hi @mkcor, I see that the CI testing is failing when trying to install scikit-image in a point before the code I added is touched.
I didn't modify anything on the setup.py or requirements that I'm aware.

I think there is an issue with the `distutils.msvccompiler` module that is being called to install the package during testing.
The same occurs trying to install scikit-image from the main branch -- I tried to install it locally and failed with the same error.

Here the error:
+ source tools/github/before_install.sh
++ set -ex
++ export PIP_DEFAULT_TIMEOUT=60
++ PIP_DEFAULT_TIMEOUT=60
++ [[ 0 == 1 ]]
++ python -m pip install --upgrade 'pip<22.1' wheel setuptools
Collecting pip<22.1
  Downloading pip-22.0.4-py3-none-any.whl (2.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 91.7 MB/s eta 0:00:00
Collecting wheel
  Downloading wheel-0.37.1-py2.py3-none-any.whl (35 kB)
Requirement already satisfied: setuptools in /opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages (56.0.0)
Collecting setuptools
  Downloading setuptools-65.0.0-py3-none-any.whl (1.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 99.1 MB/s eta 0:00:00
Installing collected packages: wheel, setuptools, pip
  Attempting uninstall: setuptools
    Found existing installation: setuptools 56.0.0
    Uninstalling setuptools-56.0.0:
      Successfully uninstalled setuptools-56.0.0
  Attempting uninstall: pip
    Found existing installation: pip 22.2.2
    Uninstalling pip-22.2.2:
      Successfully uninstalled pip-22.2.2
Successfully installed pip-22.0.4 setuptools-65.0.0 wheel-0.37.1
++ python -m pip install -r requirements/build.txt
Collecting Cython!=0.29.18,>=0.29.21
  Downloading Cython-0.29.32-cp38-cp38-manylinux_2_17_x86_64.manylinux[20](https://github.com/scikit-image/scikit-image/pull/6462/checks#step:4:21)14_x86_64.manylinux_2_24_x86_64.whl (1.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 60.5 MB/s eta 0:00:00
Collecting packaging>=20.0
  Downloading packaging-[21](https://github.com/scikit-image/scikit-image/pull/6462/checks#step:4:22).3-py3-none-any.whl (40 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.8/40.8 KB 14.9 MB/s eta 0:00:00
Collecting pythran
  Downloading pythran-0.11.0-py3-none-any.whl (4.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.2/4.2 MB 103.6 MB/s eta 0:00:00
Requirement already satisfied: wheel in /opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages (from -r requirements/build.txt (line 4)) (0.37.1)
Collecting numpy>=1.19
  Downloading numpy-1.23.2-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.1/17.1 MB 91.7 MB/s eta 0:00:00
Collecting pyparsing!=3.0.5,>=2.0.2
  Downloading pyparsing-3.0.9-py3-none-any.whl (98 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.3/98.3 KB 34.2 MB/s eta 0:00:00
Collecting beniget~=0.4.0
  Downloading beniget-0.4.1-py3-none-any.whl (9.4 kB)
Collecting gast~=0.5.0
  Downloading gast-0.5.3-py3-none-any.whl (19 kB)
Collecting ply>=3.4
  Downloading ply-3.11-py2.py3-none-any.whl (49 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 KB 19.3 MB/s eta 0:00:00
Installing collected packages: ply, pyparsing, numpy, gast, Cython, packaging, beniget, pythran
Successfully installed Cython-0.29.32 beniget-0.4.1 gast-0.5.3 numpy-1.23.2 packaging-21.3 ply-3.11 pyparsing-3.0.9 pythran-0.11.0
WARNING: You are using pip version [22](https://github.com/scikit-image/scikit-image/pull/6462/checks#step:4:23).0.4; however, version 22.2.2 is available.
You should consider upgrading via the '/opt/hostedtoolcache/Python/3.8.13/x64/bin/python -m pip install --upgrade pip' command.
++ python -m pip install -r requirements/default.txt
Requirement already satisfied: numpy>=1.19 in /opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages (from -r requirements/default.txt (line 1)) (1.[23](https://github.com/scikit-image/scikit-image/pull/6462/checks#step:4:24).2)
Collecting scipy>=1.5
  Downloading scipy-1.9.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (43.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.4/43.4 MB 53.5 MB/s eta 0:00:00
Collecting networkx>=2.5
  Downloading networkx-2.8.5-py3-none-any.whl (2.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 102.9 MB/s eta 0:00:00
Collecting pillow>=9.0.1
  Downloading Pillow-9.2.0-cp38-cp38-manylinux_2_28_x86_64.whl (3.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 101.6 MB/s eta 0:00:00
Collecting imageio>=2.4.1
  Downloading imageio-2.21.1-py3-none-any.whl (3.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 106.8 MB/s eta 0:00:00
Collecting tifffile>=2019.7.[26](https://github.com/scikit-image/scikit-image/pull/6462/checks#step:4:27)
  Downloading tifffile-2022.8.12-py3-none-any.whl (208 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 208.5/208.5 KB 59.6 MB/s eta 0:00:00
Collecting PyWavelets>=1.1.1
  Downloading PyWavelets-1.3.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 113.8 MB/s eta 0:00:00
Requirement already satisfied: packaging>=20.0 in /opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages (from -r requirements/default.txt (line 8)) (21.3)
Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages (from packaging>=20.0->-r requirements/default.txt (line 8)) (3.0.9)
Installing collected packages: tifffile, scipy, PyWavelets, pillow, networkx, imageio
Successfully installed PyWavelets-1.3.0 imageio-2.21.1 networkx-2.8.5 pillow-9.2.0 scipy-1.9.0 tifffile-2022.8.12
WARNING: You are using pip version 22.0.4; however, version 22.2.2 is available.
You should consider upgrading via the '/opt/hostedtoolcache/Python/3.8.13/x64/bin/python -m pip install --upgrade pip' command.
++ python -m pip list
Package    Version
---------- ---------
beniget    0.4.1
Cython     0.29.32
gast       0.5.3
imageio    2.21.1
networkx   2.8.5
numpy      1.23.2
packaging  21.3
Pillow     9.2.0
pip        22.0.4
ply        3.11
pyparsing  3.0.9
pythran    0.11.0
PyWavelets 1.3.0
scipy      1.9.0
setuptools 65.0.0
tifffile   2022.8.12
wheel      0.37.1
WARNING: You are using pip version 22.0.4; however, version 22.2.2 is available.
You should consider upgrading via the '/opt/hostedtoolcache/Python/3.8.13/x64/bin/python -m pip install --upgrade pip' command.
++ export -f section
++ export -f retry
++ set +ex
+ python setup.py sdist
setup.py:9: DeprecationWarning: 

  `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
  of the deprecation of `distutils` itself. It will be removed for
  Python >= 3.12. For older Python versions it will remain present.
  It is recommended to use `setuptools < 60.0` for those Python versions.
  For more details, see:
    https://numpy.org/devdocs/reference/distutils_status_migration.html 


  from numpy.distutils.command.build_ext import build_ext as npy_build_ext
Traceback (most recent call last):
  File "setup.py", line 9, in <module>
    from numpy.distutils.command.build_ext import build_ext as npy_build_ext
  File "/opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages/numpy/distutils/command/build_ext.py", line 16, in <module>
    from numpy.distutils.system_info import combine_paths
  File "/opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages/numpy/distutils/system_info.py", line 196, in <module>
    from numpy.distutils.command.config import config as cmd_config
  File "/opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages/numpy/distutils/command/config.py", line 19, in <module>
    from numpy.distutils.mingw32ccompiler import generate_manifest
  File "/opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages/numpy/distutils/mingw32ccompiler.py", line [28](https://github.com/scikit-image/scikit-image/pull/6462/checks#step:4:29), in <module>
    from distutils.msvccompiler import get_build_version as get_build_msvc_version
ModuleNotFoundError: No module named 'distutils.msvccompiler'
Error: Process completed with exit code 1.

mkcor commented:
Hi @fercer,

You are right, the CI failure is unrelated to your PR!

It's a much broader issue, reported in several places, notably: https://github.com/pyodide/pyodide/issues/2971

/cc @grlee77 @alexdesiqueira @lagru

mkcor commented:
@fercer meanwhile would you mind fixing the PEP 8 issues (as reported by pep8speaks)? Thanks :pray:

mkcor commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1777 @@
+
+            # Determine if the intersection is inside the current edge or not.
+            if is_ray:
+                inter_in_edge = tolerance <= t_coefs[0] <= 1.0-tolerance
```
```suggestion
                inter_in_edge = tolerance <= t_coefs[0] <= 1.0 - tolerance
```

mkcor commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1777 @@
+            if is_ray:
+                inter_in_edge = tolerance <= t_coefs[0] <= 1.0-tolerance
+            else:
+                inter_in_edge = tolerance <= t_coefs[0] <= 1.0-tolerance \
```
```suggestion
                inter_in_edge = tolerance <= t_coefs[0] <= 1.0 - tolerance \
```

mkcor commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1777 @@
+            else:
+                inter_in_edge = tolerance <= t_coefs[0] <= 1.0-tolerance \
+                                and \
+                                tolerance <= t_coefs[1] <= 1.0-tolerance
```
```suggestion
                                tolerance <= t_coefs[1] <= 1.0 - tolerance
```

mkcor committed:
Fix PEP 8

mkcor commented:
Dear @fercer,

I just committed some trivial fixes (PEP 8 issues), so you'll have to sync your local branch with a `git pull origin selfoverlap-polys` before you resume working on this PR.

I wanted to trigger a CI build to see if the `distutils.msvc` issue has gone (cf. https://github.com/pypa/setuptools/pull/3505).

fercer committed:
Fixing PEP 8 issues on spaces around arithmetic operations

fercer committed:
Fixed PEP 8 issues for hanging indent on a conditional continuation line

fercer committed:
Fixing PEP 8 issue with a hanging indent around a boolean operation

fercer committed:
Fixing PEP 8 issues for breaking line after boolean operation

lagru commented:
Hey @fercer, thank you very much for this large but promising contribution. This PR already looks to be of high-quality and I'm really into the thorough documentation of the private helper functions! Before I really dive into this further I'd like to get some questions out of the way, so that we can decide whether this is a good fit for scikit-image (I hope so!).

Are we allowed to provide the submitted code under scikit-image's license (BSD-3-Clause)? Is this code patented? Who is the copyright holder?

Just in general I'm really interested in more context on why this function was implemented, where it might be useful and how you use it. In scikit-image we generally aim to include well used and cited functionality. This is not a hard requirement but helps in justifying a large maintenance burden.  

Also this single function provides a huge code base. This might be due to the complexity of the solved problem but I'm really hoping that there are some opportunities to decrease the maintenance burden a little bit. But that's a task for the actual review. :wink:

I've taken a quick glance at the paper and your gallery example and I'd like you to confirm and help my understanding. So the function divides self-overlapping polygons into non-overlapping polygon. Do these sub-polygons overlap each other tho? From the [second plot](https://output.circle-artifacts.com/output/job/e8af89bf-acab-4446-b256-e63e6f3ff102/artifacts/0/doc/build/html/auto_examples/edges/plot_self_overlapping_polygon.html) I can't really confirm whether that is the case (something that could be communicated more clearly by separating the sub-polygons from each other like the paper does).

fercer committed:
Fixed gallery example where sub titles overlapped

mkcor commented:
@fercer @lagru in my understanding, this is useful when computing the area of a manifold, ensuring that certain parts (the self-overlapping ones) don't count twice. To make the gallery example more educational, I would describe the case of a folded sheet of paper. When plotting, I would use three colors: one for the background, another one for the non self-overlapping parts of the object (where there is "one layer" so to speak), and a third one for the self-overlapping parts ("two layers").

I guess the overlapping can be double ("three layers"), triple ("four layers"), ...?

fercer commented:
Hey @lagru, this is a work done at The Jackson Laboratory, and therefore it the copyright holder. However, the code is not patented and can be submitted under scikit-image's license. I'll update the `LICENSE.txt` file to state that.

The context is that at The Jackson Laboratory (JAX) we needed to generate ground-truth masks from a set of self-overlapping segmentation annotations (polygons) for a machine learning project. The issue is that when drawing some of these annotations, the overlapped regions were considered as holes because the _even-odd rule_. This happens with polygons in general.
Because we didn't find any implementation that could handle this problem, we implemented it and though that other groups could have the same issue.

As @mkcor mentioned, this is useful not only for drawing self-overlapping polygons, but also for further analysis, such is area computation, or analysis of the morphology of the resulting sub polygons. I'll update also the gallery example to include your recommendations.

The resulting sub polygons do not self-overlap with themselves but can overlap other sub polygons. However, because each sub polygon is a separate object, those overlaps can be identified now. And yes, the algorithm can handle multiple overlap layers. Additionally, the algorithm do not handle folds because it assumes that the polygon has only one _face_ up, like a highway for example.

I'm also sorry for the code base length. This is a first implementation of the algorithm in python and we didn't try to optimize prematurely. For sure there will be ways to reduce the maintenance burden in the future.

fercer committed:
Updated LICENSE.txt to add copyright of this code to The Jackson Laboratory, and updated gallery example

lagru commented:
Ah, thank you for the context and explanation! I'm starting to get a clearer picture now. To me this looks like a useful addition to scikit-image. :+1: 

> The resulting sub polygons do not self-overlap with themselves but can overlap other sub polygons. However, because each sub polygon is a separate object, those overlaps can be identified now.

So summing the area of all sub-polygons gives the true area of the self-overlapping non-folded polygon. I was asking myself by what rule sub-polygons are separated. In the [gallery example](https://output.circle-artifacts.com/output/job/cf206738-399e-4839-a6cc-70826d0f281e/artifacts/0/doc/build/html/auto_examples/edges/plot_self_overlapping_polygon.html) division happens along horizontal cuts. This is because the polygon is divided along horizontal cuts at the height of crest points as documented in the [paper](https://www.sciencedirect.com/science/article/pii/S0010448513001772?via%3Dihub), correct?

lagru commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1776 @@
+
+    Returns
+    -------
+    sub_polys : list
+        A list of ndarrays with the coordinates of the non
+        self-overlapping polygons obtained from dividing the original polygon.
```
```suggestion
    sub_polys : list of ndarray
        A list of the coordinates of the non self-overlapping polygons obtained
        from dividing the original polygon.
```

lagru commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1776 @@
+    -------
+    is_clockwise : bool or None
+        Whether the polygon is in clockwise direction or not.
+        This is None when no crest points were detected.
```
With the current implementation I don't see a way for this function to return `None`. Am I missing something?

lagru commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1776 @@
+    """
+    # Change the order of the axis to x, y from rr, cc.
+    vertices = coords[:, [1, 0]]
+    max_crest_ids, min_crest_ids, max_crest, min_crest = \
+        _get_crest_ids(vertices, tolerance=2 * np.finfo(np.float32).eps)
```
This is duplicate with the body of the `if not is_clockwise` clause. I think it can be moved after that clause:

```python
if not is_clockwise:
    vertices = vertices[::-1, :]

max_crest_ids, min_crest_ids, max_crest, min_crest = _get_crest_ids(
    vertices, tolerance=2 * np.finfo(np.float32).eps
)
```

lagru commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1776 @@
+    # Check if the polygon vertices are given in counter-clockwise direction
+    is_clockwise = _check_clockwise(vertices)
+
+    if is_clockwise is None:
+        # If the polygon does not have any crest point, it is because
+        # it is not self-overlapping.
+        return [vertices[:, [1, 0]]]
```
This is never true I think. Or at least currently the tests stay green if this if-clause is removed.

lagru reviewed (commented):
I took a first closer look. There's still a lot that I don't understand or haven't yet looked at but for now but there are already some quick suggestions from my side.

fercer commented on doc/examples/edges/plot_self_overlapping_polygon.py:
```diff
@@ -0,0 +1,94 @@
+    cc = cc - np.mean(cc).astype(np.int64) + 256 + (s % 3) * 512
+    img_subpolys_exp[rr, cc, :] = poly_col
+
+fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(nrows=2, ncols=2, figsize=(9, 4))
```
Hey @lagru! thanks for taking the time to review this implementation. I'll make the figure bigger so the titles do not overlap the figures of the first row.

lagru replied:
[Looks perfect now! Thanks.](https://output.circle-artifacts.com/output/job/081ecc66-0dfa-4b43-88aa-9368baaf7d18/artifacts/0/doc/build/html/auto_examples/edges/plot_self_overlapping_polygon.html)

fercer commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1776 @@
+    return vert_info
+
+
+def divide_selfoverlapping(coords):
```
That makes a lot of sense, we certainly don't want to make it confusing for the users when deciding what function to use. I'll include this change in the next commit.

fercer commented on skimage/measure/__init__.py:
```diff
@@ -13,6 +13,7 @@
 from ._label import label
 from .entropy import shannon_entropy
 from ._blur_effect import blur_effect
+from .sopoly import divide_selfoverlapping
```
That is correct, we added only one exposed function from that submodule and the rest are private helper functions. I'll include this too in the next commit.

lagru replied:
Could you still extend the module name and or add a short summary of the modules contents at the beginning? It's not critical but useful for long-term maintenance and I don't want it to be forgotten. :wink:

lagru replied:
E.g. https://github.com/scikit-image/scikit-image/blob/main/skimage/morphology/_flood_fill.py#L1-L5

fercer commented on skimage/measure/sopoly.py:
```diff
@@ -0,0 +1,1776 @@
+
+
+def divide_selfoverlapping(coords):
+    """Divide a self-overlapping polygon into non self-overlapping polygons.
+    This implements the algorithm proposed in [1].
```
I think that adding a "Notes" section can help to understand its functionality. As you mention "self-intersecting" is different from "self-overlapping". The first reference from the paper from where this function is based defines a self-overlapping curve as _a curve that can be divided by nontrivial line segments into simple curves_. While this definition is still a little confusing, I can try to elaborate in the "Notes" section.

fercer committed:
Apply suggestions from code review on improving docstrings of the visible function

Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

fercer committed:
Renamed sopoly.py to _sopoly.py and divide_selfoverlapping to separate_selfoverlapping_polygon

fercer commented:
The last commit includes most of the suggestions from @lagru. I decided to keep the name of the function as `separate_selfoverlapping_polygon` to point out that this works on self-overlapping polygons, and not on arbitrary overlapping polygons. I hope that makes sense when users look for this function.

Additionally, I noticed that there is one test failing. However, I believe that is on a submodule that is independent from this implementation. 

Here is the summary of the failing test.

=========================== short test summary info ============================
FAILED scikit-image/skimage/future/graph/rag.py::skimage.future.graph.rag.show_rag
===== 1 failed, 8232 passed, 15 skipped, 21 warnings in 301.30s (0:05:01) ======
Error: Process completed with exit code 1.

alexdesiqueira commented:
Thank you @fercer! 

> Additionally, I noticed that there is one test failing. However, I believe that is on a submodule that is independent from this implementation.

I can confirm, it is not related. The warning breaking that error follows:

```
=================================== FAILURES ===================================
_________________ [doctest] skimage.future.graph.rag.show_rag __________________
489     Examples
490     --------
491     >>> from skimage import data, segmentation
492     >>> from skimage.future import graph
493     >>> import matplotlib.pyplot as plt
494     >>>
495     >>> img = data.coffee()
496     >>> labels = segmentation.slic(img)
497     >>> g =  graph.rag_mean_color(img, labels)
498     >>> lc = graph.show_rag(labels, g, img)
UNEXPECTED EXCEPTION: PendingDeprecationWarning('The _get_cmap function will be deprecated in a future version. Use ``matplotlib.colormaps[name]`` instead.')
```

Feel free to go ahead with this one in the meantime, @lagru.

lagru commented on skimage/measure/_sopoly.py:
```diff
@@ -0,0 +1,1772 @@
+    is_clockwise : bool
+        Whether the polygon is in clockwise direction or not.
+    """
+    signed_area = (np.sum(vertices[:-1, 0] * vertices[1:, 1]
+                          - vertices[:-1, 1] * vertices[1:, 0])
+                   + vertices[-1, 0] * vertices[0, 1]
+                   - vertices[-1, 1] * vertices[0, 0])
+    is_clockwise = signed_area < 0
+    return is_clockwise
```
Does this solution work for all polygons? Is it the same idea as the solution presented [here](https://stackoverflow.com/a/10298685) and in [Polygon Area - Wolfram MathWorld](https://mathworld.wolfram.com/PolygonArea.html). In that case could we use [numpy.linalg.det](https://numpy.org/doc/stable/reference/generated/numpy.linalg.det.html) to simplify the calculation of the determinant? Would also be useful down the line to include the reference in the docstring.

MathWorld also states that it only works for "non-self-intersecting" polygons. Why is that not a problem here?

fercer replied:
Yes, computation of the signed area of a polygon is based on the Wolfram MathWorld solution.

Unfortunately `numy.linalg.det` works for square matrices. Because the vertices coordinates matrix is rectangular it would required to be reshaped and each row duplicated in order to use the determinant approach.

The solution from MathWorld works here because selfoverlapping polygons, while being self intersecting, does not contain _twists_ by definition. A twisted polygon will have two faces pointing upwards instead of single one. One of the faces with clockwise direction and the other in counterclockwise, causing the signed areas to substract instead of add, regardless of their sign. That could cause the area of a twisted polygon to be _0_ even when it clearly has area _greater than 0_.

lagru commented on skimage/measure/_sopoly.py:
```diff
@@ -0,0 +1,1772 @@
+    return max_crest_ids, min_crest_ids, max_crest, min_crest
+
+
+def _check_clockwise(vertices):
```
How about turning this into `def _signed_polygon_area(vertices)`? That would potentially make that function more general and potentially useful elsewhere. 

To test for clockwise ordering, we could just use `is_clockwise = _signed_polygon_area(vertices) < 0` which is still quite explicit.

fercer replied:
Thanks for the suggestion @lagru! I modified that function's name and its docstrings to show its use case.

fercer committed:
Changed names of functions and _sopoly.py file to ease maintenance burden, and added notes and examples

fercer committed:
Fixed example in docstring of function separate_selfoverlapping_polygon by adding correct indentation symbols

fercer commented:
Hey @lagru! I added some notes and examples to the main exposed function `separate_selfoverlapping_polygon` in this commit (that I forgot to add last time). I also changed the name of the file from `_sopoly.py` to `_selfoverlapping_polygon.py` and included some details of the idea behind the implemented algorithm to make it easier to maintain.

I see that some tests are failing again, but it looks like it comes from an example of a different submodule that I didn't touch. The error is related to the use of a function that comes in _matplotlib>=3.5_, and the version installed to run the tests is _matplolib==3.3_ (from the log).

Here the summary of the error
=================================== FAILURES ===================================
_________________ [doctest] skimage.future.graph.rag.show_rag __________________
200     Examples
201     --------
202     >>> from skimage import data, segmentation
203     >>> from skimage.future import graph
204     >>> import matplotlib.pyplot as plt
205     >>>
206     >>> img = data.coffee()
207     >>> labels = segmentation.slic(img)
208     >>> g =  graph.rag_mean_color(img, labels)
209     >>> lc = graph.show_rag(labels, g, img)
UNEXPECTED EXCEPTION: ImportError('"<function show_rag at 0x7f0a7c1821f0>" in "skimage.future.graph.rag" requires "matplotlib >=3.5"')

lagru commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+        The positional index of the minimum crest vertex.
+    """
+    # Get the direction of the polygon when traversing its perimeter.
+    edges = vertices[1:, :2] - vertices[:-1, :2]
```
Why the slicing with `:2` in the second dimension here? Shouldn't `vertices` have the shape `(N, 2)` at this point?

fercer replied:
That function is reused later to process an extended array that contains the vertices of the intersection cuts. That array has more columns with the extra information. That is why vertices can be (N, 2) or (N, 5). I'll add that in the parameters section of the docstring of that helper function.

lagru commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+
+    The separation algorithm implemented in this function is based on the idea
+    that a complex curve can be compressed, without twisting it, into a two
+    dimensional disk. Therefore, a set of non trivial chords (cuts) separate
```
```suggestion
    dimensional disk. Therefore, a set of non-trivial chords (cuts) separate
```

lagru commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+    that a complex curve can be compressed, without twisting it, into a two
+    dimensional disk. Therefore, a set of non trivial chords (cuts) separate
+    the disk into mutually exclusive segments [1]_. These segments, in the
+    original complex curve, are the set of resulting non selfoverlapping sub
```
```suggestion
    original complex curve, are the set of resulting non-selfoverlapping sub
```

lagru commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+    max_crest : int
+        The positional index of the maximum crest vertex.
+    min_crest : int
+        The positional index of the minimum crest vertex.
```
I think a short explanation / definition of what "crest points" are would be helpful in the long run.

fercer replied:
Yes, I'll add a definition for what crest points are detected. According to the algorithm, only crest points at _left turns_ are considered. I'll also add what a _left turn_ is for more clarity.

lagru commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+    edges = np.vstack((vertices[0, :2] - vertices[-1, :2], edges))
+    edges = edges / np.linalg.norm(edges, axis=1).reshape(-1, 1)
+
+    norm_left = np.hstack((-edges[:, np.newaxis, 1], edges[:, np.newaxis, 0]))
+
+    cos_angle = np.sum(norm_left[:-1] * edges[1:], axis=1)
+    cos_angle = np.append(cos_angle, np.sum(norm_left[-1] * edges[0]))
+
+    dir_left = cos_angle > 0.0
```
I don't understand how this determines the actual angle between two edges (?). Is this computing only the argument to the cosinus function?

fercer replied:
That is correct, the argument of the cosine function is enough there to determine if the direction is from right to left. I'll change the name of the variable and add a comment to avoid confusion.

lagru reviewed (commented):
Hey @fercer, sorry for the delay. Honestly, I find the code very hard to follow even when using a debugger even though it contains a lot of comments already. This may just speak to my current skill level but it's a barrier nevertheless. I have a feeling that a lot of the vectorized code would be actually more readable if it was using loops. I'm not suggesting this as a serious solution but I'd like to thing that for loops using Numba could be a good alternative here.

lagru commented:
See also https://discuss.scientific-python.org/t/is-the-algorithm-on-division-of-self-overlapping-polygons-a-good-fit/501.

@fercer, sorry for raising that discussion this late!

jni commented:
I just want to lend my support to including this. In general working with polygons in Python is a pain and it's hard to find good implementations, so I'm glad if we can fill that hole. (Pun not intended, but I'll take it. 😂) 🙏 Thanks @fercer and @lagru!

fercer commented on LICENSE.txt:
```diff
@@ -69,6 +69,10 @@ Function: skimage/_shared/version_requirements.py:is_installed
 Copyright: 2009-2011 Pierre Raybaut
 License: MIT
 
+Files: skimage/measure/sopoly.py
```
Thanks @lagru for reviewing my PR again and pointing out these changes.
I'll add this change to the LICENSE.txt file

fercer commented on skimage/measure/tests/test_sopoly.py:
```diff
@@ -0,0 +1,122 @@
+import numpy as np
```
Thanks for noticing that numpy was missing in the tests submodule

fercer commented:
Thanks for raising the discussion @lagru!
I understand your concern about the current state of the code.
I'd like to wait for the opinion of the community about merging this PR before moving the code from vectorized operations to loop operations (something that I'm totally fine with doing it :smile:).
I definitely want the code to be understandable and maintainable.

mkcor commented on doc/examples/edges/plot_self_overlapping_polygon.py:
```diff
@@ -0,0 +1,60 @@
+"""
+==================================
```
Thanks! It would be good to trigger CI again, so I can read the latest version of the gallery example (I'll try and make a few suggestions while reviewing, so you can commit them, which will trigger CI).

mkcor commented on doc/examples/edges/plot_self_overlapping_polygon.py:
```diff
@@ -0,0 +1,94 @@
+===================================
+
+This example shows how to separate a self-overlapping polygon into non
+self-overlapping sub polygons using the method [1]_.
```
```suggestion
self-overlapping sub-polygons using the method proposed in [1]_.
```

mkcor commented on doc/examples/edges/plot_self_overlapping_polygon.py:
```diff
@@ -0,0 +1,94 @@
+
+This example shows how to separate a self-overlapping polygon into non
+self-overlapping sub polygons using the method [1]_.
+Each sub polygon can then be drawn or analyzed separately.
```
```suggestion
Each sub-polygon can then be drawn or analyzed separately.
```

mkcor commented on skimage/measure/tests/test_sopoly.py:
```diff
@@ -0,0 +1,122 @@
+from skimage._shared.testing import assert_array_equal
+
+
+# A self-overlapping polygon that can be divided into three
+# non self-overlapping polygons.
```
```suggestion
# This is a self-overlapping polygon which can be divided into three
# non self-overlapping sub-polygons.
```

mkcor commented on skimage/measure/tests/test_sopoly.py:
```diff
@@ -0,0 +1,122 @@
+
+# A self-overlapping polygon that can be divided into three
+# non self-overlapping polygons.
+sample_cw_overlap = np.array([[200, 271], [251, 267], [312, 267], [381, 269],
```
I think we don't need the `sample_` prefix in these names...

mkcor commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+"""_selfoverlapping_polygon.py - a selfoverlapping polygon separation algorithm
+
+This module provides a function to separate a selfoverlapping polygon.
```
```suggestion
This module provides a function to separate a self-overlapping polygon.
```

mkcor commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+"""_selfoverlapping_polygon.py - a selfoverlapping polygon separation algorithm
+
+This module provides a function to separate a selfoverlapping polygon.
+Selfoverlapping polygons are defined as curves that can be obtained from
```
```suggestion
Self-overlapping polygons are defined as curves which can be obtained from
```

mkcor commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+
+This module provides a function to separate a selfoverlapping polygon.
+Selfoverlapping polygons are defined as curves that can be obtained from
+stretching a two dimensional disk without twisting it. Therefore, not all self
```
```suggestion
stretching a two-dimensional disk without twisting it. Therefore, not all self-
```

mkcor commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+This module provides a function to separate a selfoverlapping polygon.
+Selfoverlapping polygons are defined as curves that can be obtained from
+stretching a two dimensional disk without twisting it. Therefore, not all self
+intersecting polygons are selfoverlapping, even when they contain overlapping
```
```suggestion
intersecting polygons are self-overlapping, even when they contain overlapping
```

mkcor commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+intersecting polygons are selfoverlapping, even when they contain overlapping
+sections.
+
+An example of a selfintesecting polygon that is not selfoverlapping is a sheet
```
```suggestion
An example of a self-intersecting polygon that is not self-overlapping is a sheet
```

mkcor commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+sections.
+
+An example of a selfintesecting polygon that is not selfoverlapping is a sheet
+of paper that has been folded at least one time. That would casue to have two
```
```suggestion
of paper that has been folded at least one time. That would cause two
```

mkcor commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+
+An example of a selfintesecting polygon that is not selfoverlapping is a sheet
+of paper that has been folded at least one time. That would casue to have two
+faces of the same sheet pointing upwards. On the other hand, an example of a
```
```suggestion
faces of the same sheet to point upwards. Conversely, an example of a
```

mkcor commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+An example of a selfintesecting polygon that is not selfoverlapping is a sheet
+of paper that has been folded at least one time. That would casue to have two
+faces of the same sheet pointing upwards. On the other hand, an example of a
+selfoverlapping polygon is the loop formed by an exit ramp on a highway.
```
```suggestion
self-overlapping polygon is the loop formed by an exit ramp on a highway.
```

mkcor commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+faces of the same sheet pointing upwards. On the other hand, an example of a
+selfoverlapping polygon is the loop formed by an exit ramp on a highway.
+
+This function is an implementation of the method proposed in
```
```suggestion
This code is an implementation of the method proposed in
```

mkcor commented on skimage/measure/_selfoverlapping_polygon.py:
```diff
@@ -0,0 +1,1875 @@
+Mukherjee, "Self-overlapping curves: Analysis and applications", Computer-Aided
+Design 46 (2014) 227-232.
+"""
+import numpy as np
+import math
+from functools import reduce
```
```suggestion
from functools import reduce
import math
import numpy as np
```

lagru commented:
> I'd like to wait for the opinion of the community about merging this PR before moving the code from vectorized operations to loop operations (something that I'm totally fine with doing it smile).

Fyi, as I just noticed, we can use Pythran in our library. If we decide to switch some stuff from vectorized to loops that might be easier than using Cython if your are not that familiar with it.

fercer committed:
Apply suggestions from code review to improve docstrings

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>
Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

mkcor commented:
Hi @fercer,

Thank you for improving the gallery example so much! I believe that we need to start by defining what we mean by 'self-overlapping polygon' (even if we point to a reference). The abstract of the reference in question starts with: "When a disk in 2D is stretched arbitrarily with possible self-overlaps, without twisting it, its boundary forms a complex curve known as a self-overlapping curve." So, typically, it's worth specifying if/that 'self-overlapping polygon' refers to the surface enclosed by such a curve(?).

mkcor commented:
> "When a disk in 2D is stretched arbitrarily with possible self-overlaps, without twisting it, its boundary forms a complex curve known as a self-overlapping curve."

Also, this sentences assumes that *everybody knows what 'self-overlaps' are* (which is not my case, so probably isn't either for the average reader of the gallery): Since a folded sheet of paper *isn't* 'self-overlapping' (as you write in the header of `skimage/measure/_selfoverlapping_polygon.py`, clearly the definition of a self-overlap is neither common (as in, common language) nor trivial.

fercer committed:
Added an explanation of what is a self-overlapping polygon and a counterexample

fercer committed:
Merge changes from code review into this branch

fercer committed:
Simplified ordering of rays in _sort_rays helper function

fercer committed:
Fixed error in helper function _sort_rays

fercer commented:
Thanks for reviewing my code again @mkcor, and thanks for your comments.
I included the same description of self-overlapping polygons from _selfoverlapping_polygon.py into the gallery example.
However, do you think that I should add another plot in the same gallery example to show the difference?

fercer commented:
Hi @jorisvandenbossche and @Parcly-Taxel, I agree with moving this PR to a more polygon focused package if you are interested.
Please let me know if this functionality would fit into `Shapley` or `SymPy`.

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
